corehq/blobs/mixin.py

```diff
@@ -57,32 +76,55 @@ class Meta(object):
 
     _atomic_blobs = None
 
-    def _blobdb_bucket(self):
-        if self._id is None:
-            raise ResourceNotFound(
-                "cannot manipulate attachment on unidentified document")
-        return join(_get_couchdb_name(type(self)), safe_id(self._id))
+    @classmethod
+    def wrap(cls, data):
+        if data.get("external_blobs"):
+            doc_id = safe_id(data["_id"])
+            dbname = _get_couchdb_name(cls)
+            normalize = BlobMetaPointer._normalize_json
+            blobs = {}
+            normalized = False
+            for key, value in data["external_blobs"].items():
+                if value["doc_type"] == "BlobMetaPointer":
+                    blobs[key] = value
+                else:
+                    blobs[key] = normalize(dbname, doc_id, value)
+                    normalized = True
+            if normalized:
+                data = data.copy()
```

Comment:
why is the `copy` necessary?

Reply:
Yea I know that's generally a good idea. In this case the data is coming from an HTTP request so modifying it shouldn't really be an issue. The reason I brought it up is that it's going to double the memory requirement for any doc loaded (which in the case of largs apps could be significant).

Reply:
The other thing is that JsonObject mutates the dictionary under the hood anyway.

Reply:
It's only a shallow copy, so it will not result in that much extra memory use. Also, just because `JsonObject` does a bad thing doesn't make me feel better about doing it in other places—the hard-to-find bugs get harder to find when it's done in multiple places.